proposals/consolidate-merged-lifecycle-hook-schema.md

```diff
@@ -0,0 +1,142 @@
+# Goal
+
+This proposal aims to raise a concept of the 'mergedConfiguration' into the specification, so that implementing tools can generate, as well as consume, a 'merged' `devcontainer.json` file that is the computed merge after parsing a user 'devcontainer.json'.  [The merging logic is already documented](https://containers.dev/implementors/spec/#merge-logic). This proposal focuses on standardizing an output JSON format to represent this merge.
+
+Properties that can not or should not be represented as `devcontainer.json` properties are prepended with a `$`.  This proposal standardizes how these properties should be processed and outputted to file, so that implementing tools can write/consume them consistently.
+
+## Motivation
+
+The current generated 'mergedConfiguration' returned by the `read-configuration` CLI command does not return a `devcontainer.json`. [(visualization)](https://github.com/devcontainers/cli/pull/390#issuecomment-1430190326).  By standardizing the output format, implementing tools can generate a portable, merged `devcontainer.json` that can be consumed by other tools.
+
+## Proposal
+
+### A. Update the devcontainer.json schema
+
+#### A1. Extends lifecycle script properties
+
+Add a value (`LifecycleCommandParallel[]`) to the unioned definition of each lifecycle hook (except 'initializeCommand').  This allows for a syntax that captures the existing allowed types for a lifecycle hook and additionally provides a pattern for attaching additional metadata to a merged configuration.
+
+```typescript
+LifecycleCommand = string | string[]
+LifecycleCommandParallel = { [key: string]: LifecycleCommand; $origin?: string };
+```
+
+In a `devcontainer.json`, each of the following lifecycle hooks shown below can now be represented by one of these three types.
+
+```typescript
+{
+    // ...other devcontainer.json properties...
+	onCreateCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
```

Comment:
Oh, this change specifically affects the schema for `devcontainer.json`, I see. I hadn't realized that `devcontainer-feature.json` has a distinct json schema.

Reply:
Yes, I don't think we'd want to update the `devcontainer-feature.json` schema, which is very close to a `devcontainer.json` but distinct.  The goal of the `$

Reply:
> The goal of the `$` properties as members of the `devcontainer.json` is so that a command such as `devcontainer read-configuration` can output a json object that _is_ also a devcontainer.json. That's not the case today without these changes.

I've been thinking more about this, and I'm not sure that's a good thing to have in the spec itself...it seems like an implementation detail for the specific tool in question. The way some of the other comments read indicates that it'd be "valid" in a `devcontainer.json` but not actually used (re: a comment above about `$entrypoints`)...which kind of implies that the output of `read-configuration` is not *actually* a `devcontainer.json`.

That is, it seems like a "proper" implementation would use this schema to encode/decode metadata stored in the image label, and use it internally as the result of merging all of the various metadata sources, but it's not actually valid `devcontainer.json` config. Effectively splitting out merged metadata into a separate schema and leaving the existing `devcontainer.json` and `devcontainer-feature.json` alone.

The way I've been thinking about it lately is that a `*.json` contains two sets of data: config and metadata. I parse either `devcontainer.json` or `devcontainer-feature.json` into `Config` or `FeatureConfig` respectively, plus a `Metadata` entry. I parse the image label into a `MergedMetadata` entry. Internally I store the config and the merged metadata to drive the build process. I use the merged metadata to render into an image label. I print the config and merged metadata during debug output as two separate entities.